### PR TITLE
PositionManager invariants

### PR DESCRIPTION
**Pull request by ith-harvey**
# Description of change
## High level
* Added PositionManager invariant scaffolding
* Added asserts for PostionManger in handlers
* Added `PM1`, `PM2` & `PM3` invariant check
* Cleaned up invariant logging

### CODE REVIEW

File: tests/forge/invariants/PositionsAndRewards/RewardsInvariants.t.sol [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    function setUp() public override virtual {
+
+        super.setUp();
+
+        _rewards = new RewardsManager(address(_ajna), _position);
+
+        excludeContract(address(_positionHandler));
+        excludeContract(address(_rewards));
+
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/RewardsInvariants.t.sol:27]: Should also use target selector similar to GrantFund in order to prevent some of the calls in the depth specified from going to view functions that can't change checked state: 
```
        // set the list of function selectors to run
        bytes4[] memory selectors = new bytes4[](2);
        selectors[0] = _standardHandler.fundingVote.selector;
        selectors[1] = _standardHandler.updateSlate.selector;

        // ensure utility functions are excluded from the invariant runs
        targetSelector(FuzzSelector({
            addr: address(_standardHandler),
            selectors: selectors
        }));
```

Comment (MikeHathaway) [reply tests/forge/invariants/PositionsAndRewards/RewardsInvariants.t.sol:27]: Per discussion in Discord this is handled by usage of exclude contract. Changes to this are being disused out of scope

File: tests/forge/invariants/PositionsAndRewards/RewardsInvariants.t.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        // get rewards that have been claimed
+        uint256 claimedRewards  = IBaseHandler(_handler).totalRewardPerEpoch(curEpoch);
+
+        // total ajna burned by the pool over the epoch
+        (, uint256 totalBurnedInPeriod,) = _getPoolAccumulators(address(_pool), curEpoch + 1, curEpoch);
+
+        uint256 rewardsCap = Maths.wmul(totalBurnedInPeriod, 0.1 * 1e18);
+
+        // check claimed rewards < rewards cap
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/RewardsInvariants.t.sol:77]: require statements are recommended by Forge team due to some edge case bugs with assert. Also provides nice error messages in stack

File: tests/forge/invariants/PositionsAndRewards/PositionsInvariants.t.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            uint256[] memory tokenIds = IBaseHandler(_handler).getTokenIdsByBucketIndex(bucketIndex);
+            for (uint256 k = 0; k < tokenIds.length; k++) {
+                uint256 tokenId = tokenIds[k];
+                
+                (uint256 posLp, uint256 posDepositTime) = _position.getPositionInfo(tokenId, bucketIndex);
+                posLpAccum += posLp;
+                mostRecentDepositTime = (posDepositTime > mostRecentDepositTime) ? posDepositTime : mostRecentDepositTime;
+            }
+
+            assertEq(poolLpAccum, posLpAccum, "Positions Invariant PM1");
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/PositionsInvariants.t.sol]: Recommend converting to `require()`

File: tests/forge/invariants/PositionsAndRewards/handlers/unbounded/BasePositionsHandler.sol [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return tokenIdsByBucketIndex[bucketIndex_].values();
+    }
+
+    function getTokenIds() public view returns(uint256[] memory) {
+        return tokenIdsMinted.values();
+    }
+
+
+
+}
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/handlers/unbounded/BasePositionsHandler.sol:65]: nit extra spacing and no new line at EOF

**Review by MikeHathaway — APPROVED** [review]
LGTM - just one recommendation to convert `assert` into `require` in the invariant files

File: tests/INVARIANTS.md [hunk 5]
```diff
@@ -74,3 +74,18 @@
...
+- **RW6**: Each time the bucket rate is updated the `updateRewardsClaimed` accumulator should be properly updated
+- **RW7**: After unstaking or claiming rewards, if ajna has been burned over an epoch while the staker was staked they should have an increased amount in ajna balance `_ajna.balanceOf()` that matches the amount of ajna balance deducted from the contract`_ajna.balanceOf()`.
+- **RW8**: After unstaking, stakeInfo's tokenId mapping should be deleted and all corrosponding values should therefore be zero'd out (`StakeInfo.ajnaPool`, `StakeInfo.lastClaimedEpoch`, `StakeInfo.owner`, `StakeInfo.stakingEpoch`, `StakeInfo.snapshots`)
+
+## Position Manager
+- **PM1**: For each position, for each bucket in the position summed LP balance == LP balance of the position manager in the underlying pool
+- **PM2**: Position deposit time (`depositTime`) tracked by tokenId in PositionManager (`positions[tokenId][index]`) should always be of equal or lesser value than the PositionManager's LP at that index in the pool contract.
```
Comment (grandizzy) [inline tests/INVARIANTS.md]: we need **PM3** here to ensure LP from a healthy bucket to a bankrupt bucket are always redeemable. e.g. from bucket is healthy, deposit time 100, to bucket bankrupt at time 50, after moving LP they should be redeemable / deposit time should be > thank 50

File: tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { Strings } from '@openzeppelin/contracts/utils/Strings.sol';
 import '@openzeppelin/contracts/utils/structs/EnumerableSet.sol';
 
 import { ERC20Pool }        from 'src/ERC20Pool.sol';
+// import { RewardsManager }   from 'src/RewardsManager.sol';
```
Comment (grandizzy) [inline tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol]: this can be removed

**Comment by prat-gpt** [comment]
[8a000bd](https://github.com/ajna-finance/contracts/pull/793/commits/8a000bd3b588ac70398af5cf4f61ca716710a489) this commit has an regression test that was failing when I was working on this PR. I think it has not been resolved. The github action is passing because position manager regressions are not included in the `make test-regression-all` command.

**Comment by prat-gpt** [comment]
> [8a000bd](https://github.com/ajna-finance/contracts/pull/793/commits/8a000bd3b588ac70398af5cf4f61ca716710a489) this commit has an regression test that was failing when I was working on this PR. I think it has not been resolved. The github action is passing because position manager regressions are not included in the `make test-regression-all` command.

PR https://github.com/ajna-finance/contracts/pull/815 will fix the failing regression test

**Comment by grandizzy** [comment]
is this still needed, there's also https://github.com/ajna-finance/contracts/pull/854

**Comment by grandizzy** [comment]
> is this still needed, there's also #854

feel free to reopen if needed, per my understanding this is covered in #854
